packages/gcf-utils/src/gcf-utils.ts

```diff
@@ -885,12 +892,59 @@ export class GCFBootstrapper {
     };
   }
 
+  private async getCloudRunUrl(
+    projectId: string,
+    location: string,
+    botName: string
+  ): Promise<string | null> {
+    // Cloud Run service names can only use dashes
+    const serviceName = botName.replace('_', '-');
+    const auth = new GoogleAuth({
+      scopes: ['https://www.googleapis.com/auth/cloud-platform'],
+    });
+    const authClient = await auth.getClient();
+    const client = await run({
+      version: 'v1',
+      auth: authClient,
+    });
+    const name = `projects/${projectId}/locations/${location}/services/${serviceName}`;
+    const res = await client.projects.locations.services.get({
+      name,
+    });
+
+    if (res.data.status?.address?.url) {
```

Comment:
If undefined is ok, you could return `res.data.stats?.address?.url` here and get the same effect.

Reply:
The response field is `string | null` so it'd be weird to return `string | null | undefined`. I'd prefer to limit to one missing type but maybe return null and undefined is preferred?